### ISSUE TITLE
Update build.cake to check version of dotnet

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#2111]
The most recent release of `dotnet` CLI does not work with our cake build script unless the correct SDK is installed. Several users have run into problems building because of this. We should check that the correct version of the SDK is installed, not just that the CLI is installed.

My email response to one is below,

---

The .NET Core code was developed for the older .NET Core tooling. The initial failure is restoring NuGet packages.

You are having the issue because you do not have .NET Core `1.0.0-preview2-003131` installed (see global.json).

At the command line, if you type `dotnet --version` in the NUnit directory, you should see 1.0.0-preview2-003131 because .NET Core tries to respect the global.json. You will likely see just 1.0.0.

For example, when I type `dotnet --info` another directory, I see,

```
D:\src
λ dotnet --info
.NET Command Line Tools (1.0.0)

Product Information:
 Version:            1.0.0
 Commit SHA-1 hash:  e53429feb4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15031
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.0
```

But when I type it in the NUnit source directory, I see

```
D:\src\NUnit\nunit [master ≡]
λ dotnet --info
.NET Command Line Tools (1.0.0-preview2-003131)

Product Information:
 Version:            1.0.0-preview2-003131
 Commit SHA-1 hash:  635cf40e58

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15031
 OS Platform: Windows
 RID:         win10-x64
```

I believe that you can download that version of .NET Core from https://github.com/dotnet/core/blob/master/release-notes/download-archives/1.0.1-preview2-download.md

**Comment by ChrisMaddock** [comment]
@rprouse - Just to double check, as I'm unfamiliar here - would it also be sensible to update to sdk version 1.0.1, rather than a preview version at the same time?

I wanted to try and get the .NET Standard builds running on Travis at some point, I may revisit that soon.

**Comment by xplicit** [comment]
I tried to run build.cmd and netstandard build fails with messages "can not find type or namespace":

`dotnet --version` shows 1.0.2 in nunit folder despite of `global.json` with 

    "sdk": {
        "version": "1.0.0-preview2-003131"
    }


```
========================================
BuildNetStandard
========================================
Executing task: BuildNetStandard
Microsoft (R) Build Engine версии 14.0.25420.1
(C) Корпорация Майкрософт (Microsoft Corporation). Все права защищены.

C:\Projects\nunit\src\CommonAssemblyInfo.cs(29,12): error CS0246: Не удалось найти тип или имя пространства имен "AssemblyCompanyAttribute" (возможно, отсутствует директива using или ссылка на сборку). [C:\Pr
C:\Projects\nunit\src\CommonAssemblyInfo.cs(29,12): error CS0246: Не удалось найти тип или имя пространства имен "AssemblyCompany" (возможно, отсутствует директива using или ссылка на сборку). [C:\Projects\nu
C:\Projects\nunit\src\CommonAssemblyInfo.cs(30,12): error CS0246: Не удалось найти тип или имя пространства имен "AssemblyProductAttribute" (возможно, отсутствует директива using или ссылка на сборку). [C:\Pr
C:\Projects\nunit\src\CommonAssemblyInfo.cs(30,12): error CS0246: Не удалось найти тип или имя пространства имен "AssemblyProduct" (возможно, отсутствует директива using или ссылка на сборку). [C:\Projects\nu
C:\Projects\nunit\src\CommonAssemblyInfo.cs(31,12): error CS0246: Не удалось найти тип или имя пространства имен "AssemblyCopyrightAttribute" (возможно, отсутствует директива using или ссылка на сборку). [C:\
```

**Comment by ChrisMaddock** [comment]
@xplicit - have you installed specifically .NET sdk version 1.0.0-preview2-003131 from the link in the first post?

Another user reported they needed to also uninstall the later versions of the SDK - although I haven't seen that problem myself.

**Comment by xplicit** [comment]
@ChrisMaddock I've got installed before it because I ported [ServiceStack](http://github.com/ServiceStack/ServiceStack) to .NET Core, currently we are moving to new csproj format and I installed VS 2017 on the my dev machine. 

If I run `dotnet --version` from ServiceStack root folder, I get a message `1.0.0-preview2-1-003177` if I do the same from NUnit root folder, I get `1.0.2`

I am trying to build NUnit, because I've got an issues with running NUnit and NUnitLite tests (they both don't work after migrating ServiceStack project to new format), so I am trying to debug and found the cause why NUnitLite does not work (for NUnit I saw the issue that it does not support new csproj format).

**Comment by ChrisMaddock** [comment]
@xplicit - afraid you've exhausted my basic dotnet tooling knowledge - sorry! Hopefully @rprouse may have some ideas?

**Comment by rprouse** [comment]
@xplicit it is likely that you have `1.0.0-preview2-1-003177` and `1.0.2` installed, but if you look at NUnit's `global.json`, it requires `1.0.0-preview2-003131`. One preview2 won't fall back to the other. Annoying isn't it? I hope it ends up being easier with the new csproj format. I would like to port our .NET Standard projects to it, but it would require everyone to update to Visual Studio 2017.

@ChrisMaddock if you want to tackle getting .NET Standard building on Linux, take a look at https://github.com/nunit/nunit-console/blob/issue-10d/.travis/install.sh and https://github.com/nunit/nunit-console/blob/issue-10d/.travis.yml for a start. I still don't have it working, but it is close ☹️ You also need to do a git chmod on the `install.sh` so it gets the execute bit set on Linux. `git update-index --chmod=+x .travis/install.sh`

**Comment by ChrisMaddock** [comment]
I was just having a look at your fork actually @rprouse. 😄 

I have the framework build working using `dotnet: version` in the .travis.yml file, instead of installing manually. (It needs a lot of tidying up though, and there's failing tests.)

Here's my fork if you're interested - I'm currently doing mono and .NET Core tests on separate machines, but I'm not sure that's necessary anymore. travis.yml is probably interesting - and I have an alternate build.sh for when mono isn't installed. (Although that of course can be scrapped, if I can get them back on the same machine.) The changes in build.cake were mainly avoiding what I think were bugs - I need to go through now it's up and running, and see which were actually necessary. 

https://github.com/ChrisMaddock/nunit/tree/travis_dotnet

**Comment by xplicit** [comment]
@rprouse ah, you're right, I had all recent dotnet SDKs installed (-3121, -3177, 1.0.2) but did not have -3131 which is installed with .NET Core 1.0.1. By the way, windows version of SDK is named '-3133' so I had to change `global.json` content to make `dotnet --version` work in nunit directory. 

For me this dotnet versioning is frustrating and much difficult to remember. I wonder, why MS doesn't want to use simple version numbers like 1.0, 1.1 and so on instead of these `previewN-xxyy` numbering. Moreover some SDK versions for linux differs from SDK version for windows and this can produce build errors.

So finally I was able to build nunit netstandard version by installing sdk "1.0.0-preview2-003133" and changing `global.json` to this version. Changing to "1.0.0-preview2-1-3177" in global.json seems to work too.

For someone who will get similar issue and trying to understand which dotnet sdks are installed the following command lists installed dotnet sdk versions:

```
dir /b "%ProgramFiles%\dotnet\sdk"
```

**Comment by masaeedu** [comment]
@rprouse To fix this, is it sufficient to simply check whether the version of the .NET SDK on PATH exactly matches that of the `global.json`, or do you need more sophisticated version compatibility checks? I have a pretty good idea of where in the Cake build file the change needs to be made.

**Comment by mikkelbu** [comment]
I don't think that they should match exactly. I can build `BuildNetStandard13` and `BuildNetStandard16` but `dotnet --info` (in the nunit folder) gives me
```
.NET Command Line Tools (1.0.1)

Product Information:
 Version:            1.0.1
 Commit SHA-1 hash:  005db40cd1

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.1

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d
```
I had to install the `.NET Core 1.1.2 runtime`, but I cannot remember if it was when I updated VS or when I installed dotnet SDK ver 2.0.0, in order to be able to build the BuildNetStandards again.

**Comment by masaeedu** [comment]
@mikkelbu Can you get a listing of your `C:\Program Files\dotnet\sdk` folder? You can have multiple SDKs installed simultaneously, and the `global.json` dictates which one will be used (if available).

**Comment by mikkelbu** [comment]
Here you are
```
PS C:\Program Files\dotnet\sdk> ls

    Directory: C:\Program Files\dotnet\sdk

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
da----       11/04/2017     00:06                1.0.1
da----       27/09/2017     21:07                2.0.0
d-----       27/09/2017     21:07                NuGetFallbackFolder
```

**Comment by masaeedu** [comment]
Hmm. I was guessing you had the exact version installed, but looks like this isn't the case. So I guess there's no real rhyme or reason to which versions of the `dotnet` sdk are compatible and which are not: `1.0.1` seems to be good, whereas my `1.0.4` is not.

Maybe the Cake build should just ignore build failures before running the test step.

**Comment by rprouse** [comment]
The other issue we are going to run into here is that newer versions of Cake will install a specific version of .NET Core if the version it needs is not installed and put that in the path before the installed version of `dotnet`. I ran into this in the adapter and eventually gave in and fixed the adapter to compile using the version of `dotnet` that Cake installed. I did update the version in the `build.ps1` file to 1.1.4 We could do the same here - now that Cake requires .NET Core, just use the version Cake requires.

See https://github.com/nunit/nunit3-vs-adapter/pull/398
